File: server/routes/aiNews.js
// routes/aiNews.js
const express = require('express');
const router = express.Router();
const AINews = require('../models/AINews');

// GET all AI news items
router.get('/', async (req, res) => {
  try {
    const news = await AINews.find();
    res.json(news);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// GET a single AI news item by ID
router.get('/:id', async (req, res) => {
  try {
    const newsItem = await AINews.findById(req.params.id);
    if (!newsItem) return res.status(404).json({ message: 'News item not found' });
    res.json(newsItem);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// POST a new AI news item
router.post('/', async (req, res) => {
  const newsItem = new AINews(req.body);
  try {
    const newNewsItem = await newsItem.save();
    res.status(201).json(newNewsItem);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PUT update an AI news item
router.put('/:id', async (req, res) => {
  try {
    const updatedNewsItem = await AINews.findByIdAndUpdate(req.params.id, req.body, { new: true });
    if (!updatedNewsItem) return res.status(404).json({ message: 'News item not found' });
    res.json(updatedNewsItem);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// DELETE an AI news item
router.delete('/:id', async (req, res) => {
  try {
    const deletedNewsItem = await AINews.findByIdAndDelete(req.params.id);
    if (!deletedNewsItem) return res.status(404).json({ message: 'News item not found' });
    res.json({ message: 'News item deleted successfully' });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;

File: server/routes/ai_news.py
import json
from flask import Blueprint, request, jsonify
from models.ai_news import AINews

ai_news = Blueprint('ai_news', __name__)

def serialize(doc):
    """Convert a MongoEngine document (or QuerySet) to a Python dict/list."""
    return json.loads(doc.to_json())

@ai_news.route('/', methods=['GET'], strict_slashes=False)
def get_ai_news():
    try:
        # Sort by date in descending order (newest first)
        news_items = AINews.objects().order_by('-date')
        # Convert QuerySet to a JSON serializable Python object
        news_data = serialize(news_items)
        return jsonify(news_data), 200
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@ai_news.route('/<string:news_id>', methods=['GET'], strict_slashes=False)
def get_news_item(news_id):
    try:
        news_item = AINews.objects.get(id=news_id)
        return jsonify(serialize(news_item)), 200
    except AINews.DoesNotExist:
        return jsonify({'message': 'News item not found'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@ai_news.route('/', methods=['POST'], strict_slashes=False)
def create_news_item():
    data = request.get_json()
    try:
        news_item = AINews(**data).save()
        # Convert the saved document to a Python dict
        return jsonify(serialize(news_item)), 201
    except Exception as e:
        return jsonify({'message': str(e)}), 400

@ai_news.route('/<string:news_id>', methods=['PUT'], strict_slashes=False)
def update_news_item(news_id):
    data = request.get_json()
    try:
        news_item = AINews.objects.get(id=news_id)
        news_item.update(**data)
        news_item.reload()
        return jsonify(serialize(news_item)), 200
    except AINews.DoesNotExist:
        return jsonify({'message': 'News item not found'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 400

@ai_news.route('/<string:news_id>', methods=['DELETE'], strict_slashes=False)
def delete_news_item(news_id):
    try:
        news_item = AINews.objects.get(id=news_id)
        news_item.delete()
        return jsonify({'message': 'News item deleted successfully'}), 200
    except AINews.DoesNotExist:
        return jsonify({'message': 'News item not found'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 500
